### PR TITLE
Fix not rebound hooks after clone

### LINKED ISSUE #212: looks like misspelled ->table
This line possibly accesses non-existent property.

https://github.com/atk4/data/blob/develop/src/Join.php#L328

### PR DESCRIPTION
**Pull request by mvorisek**
merge together with https://github.com/atk4/core/pull/272

BC break - if you clone object with hooks bounded to a different object, it throws correctly (but less strict now because of https://github.com/atk4/core/pull/275 PR)

### CODE REVIEW

File: docs/advanced.rst [hunk 1]
```diff
@@ -652,23 +652,23 @@ Here is how to add them. First you need to create fields::
 I have declared those fields with never_persist so they will never be used by
 persistence layer to load or save anything. Next I need a beforeSave handler::
 
-    $this->onHook(Model::HOOK_BEFORE_SAVE, function($m) {
-        if($m->_isset($m['client_code') && !$m->_isset($m['client_id')) {
+    $this->onHookShort(Model::HOOK_BEFORE_SAVE, function() {
+        if($this->_isset($this['client_code') && !$this->_isset($this['client_id')) {
```
Comment (georgehristov) [inline docs/advanced.rst]: Should be `$this->_isset('client_code')` I believe

File: tests/Model/Smbo/Transfer.php [hunk 2]
```diff
@@ -23,12 +23,12 @@ protected function init(): void
 
         $this->addField('destination_account_id', ['never_persist' => true]);
 
-        $this->onHook(self::HOOK_BEFORE_SAVE, function ($m) {
+        $this->onHookShort(self::HOOK_BEFORE_SAVE, function () {
             // only for new records and when destination_account_id is set
-            if ($m->get('destination_account_id') && !$m->id) {
+            if ($this->get('destination_account_id') && !$this->id) {
```
Comment (georgehristov) [inline tests/Model/Smbo/Transfer.php]: Migrate out of $model->id?

Comment (mvorisek) [reply tests/Model/Smbo/Transfer.php]: wdym?

File: docs/advanced.rst (hunk 1 above)
Comment (mvorisek) [reply docs/advanced.rst]: yes, old typo, fixed, thanks

**Review by georgehristov — COMMENTED** [review]
Generally LGTM. I am a bit worried that it gets complicated...
Does it completely solve the issue of hook binding on cloning?

File: src/Field.php [hunk 3]
```diff
@@ -223,6 +223,21 @@ public function __construct(array $defaults = [])
...
     }
 
+    protected function onHookToOwner(string $spot, \Closure $fx, array $args = [], int $priority = 5): int
+    {
+        $name = $this->short_name;
+
+        return $this->owner->onHookDynamic(
+            $spot,
+            static function (Model $owner) use ($name) {
+                return $owner->getField($name);
```
Comment (georgehristov) [inline src/Field.php:233]: Why not `return $owner->getField($this->short_name);`

Comment (mvorisek) [reply src/Field.php:233]: because we will prevent (at least) old `$this` to be GCed, goiing to sleep today ...zzz

**Review by DarkSide666 — APPROVED** [review]
Well... OK :)

**Pull request merged**